File: part1/js/ex3b.js
// Tree a
function Tree() {
    this.node = null; // null | a
    this.left = null; // null | Tree a
    this.right = null; // null | Tree a
}

// (Ord a, Eq a) => Tree a -> a -> Tree a
function appendToTree(tree, a) {
    if(tree === null) {
        tree = new Tree();
        tree.node = a;
        return tree;
    }

    if(tree.node === null) {
        tree.node = a;
        return tree;
    }

    if(tree.node === a) {
        return tree;
    }

    if(a > tree.node) {
        tree.right = appendToTree(tree.right, a);
        return tree;
    }

    if(a < tree.node) {
        tree.left = appendToTree(tree.left, a);
        return tree;
    }
}


// (Ord a) => [a] -> Tree a
function buildTree(list) {
    var root = new Tree();
    root.node = list.shift();

    var tree = list.reduce(appendToTree, root);

    return tree;
}

// (Eq a) => Tree a -> a -> Maybe a
function findInTree(tree, item) {
    if(tree === null) {
        return null;
    }

    if(tree.node === item) {
        return item;
    }

    if(item > tree.node) {
        return findInTree(tree.right, item);
    } else {
        return findInTree(tree.left, item);
    }
}

/*
 * Implement the BinarySearch algorithm using a Binary Tree
 */
// (Ord A, Eq a) => [a] -> a -> Maybe a
function search(list, item) {
    var tree = buildTree(list);
    return findInTree(tree, item);
}

module.exports = search;

File: part1/hs/ex3b.hs
data Tree a = EmptyTree | Node a (Tree a) (Tree a) deriving (Show)

appendtotree :: (Ord a) => Tree a -> a -> Tree a
appendtotree EmptyTree x = (Node x EmptyTree EmptyTree)
appendtotree (Node a l r) x
    | x == a = (Node a l r)
    | x > a = (Node a l (appendtotree r x))
    | otherwise = (Node a (appendtotree l x) r)

buildtree :: (Ord a) => [a] -> Tree a
buildtree xs = foldl appendtotree EmptyTree xs

findintree :: (Ord a, Eq a) => a -> Tree a -> Maybe a
findintree _ EmptyTree = Nothing
findintree x (Node a l r)
    | x == a = Just x
    | x > a = findintree x r
    | otherwise = findintree x l

search :: (Ord a) => [a] -> a -> Maybe a
search xs x = findintree x $ buildtree xs
